airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py

```diff
@@ -460,12 +486,172 @@ class Audience(GoogleAdsStream):
+            sync_mode=sync_mode, cursor_field=cursor_field, stream_state=stream_state.get(self.parent_stream_name)
+        ):
+            customer_id = parent_slice.get("customer_id")
+            child_slice = {"customer_id": customer_id, "updated_ids": set(), "deleted_ids": set()}
+            if not self.current_state(customer_id):
+                yield child_slice
+                continue
+
+            parent_slice["resource_type"] = self.resource_type
+            stream_is_updated = False
+            for parent_record in self.parent_stream.read_records(sync_mode=sync_mode, cursor_field=cursor_field, stream_slice=parent_slice):
+                substream_id = parent_record.get(self.parent_id_field)
+                if not substream_id:
+                    continue
+                stream_is_updated = True
+                # Add ids to set of changed or deleted items
+                if parent_record.get("change_status.resource_status") == "REMOVED":
+                    child_slice["deleted_ids"].add(substream_id)
+                else:
+                    child_slice["updated_ids"].add(substream_id)
+
+            # update parent state in child stream
+            if stream_is_updated:
+                self._state = {self.parent_stream_name: self.parent_stream.state}
+            else:
+                # full refresh sync without parent stream
+                self._state = {
+                    self.parent_stream_name: {
+                        self.parent_cursor_field: pendulum.today().start_of("day").format("YYYY-MM-DD HH:mm:ss.SSSSSS")
+                    }
+                }
+
+            if stream_is_updated:
+                yield child_slice
+            else:
+                yield from []
+
+    def read_records(
+        self, sync_mode: SyncMode, cursor_field: List[str] = None, stream_slice: MutableMapping[str, Any] = None, **kwargs
+    ) -> Iterable[Mapping[str, Any]]:
+        """
+        This method is overridden to read records using parent stream
+        """
+        self.incremental_sieve_logger.bump()
+
+        self.incremental_sieve_logger.info(f"Started reading records for slice: {stream_slice}")
+
+        records = super().read_records(sync_mode, stream_slice=stream_slice)
+        for record in records:
+            yield record
+
+        # update parent state in child stream
+        if self.parent_stream.state:
+            self._state = {self.parent_stream_name: self.parent_stream.state}
+        else:
+            # full refresh sync without parent stream
+            self._state = {self.parent_stream_name: {self.parent_cursor_field: pendulum.today().start_of("day").format("YYYY-MM-DD")}}
+
+        # yield deleted items
+        record_fields = GoogleAds.get_fields_from_schema(self.get_json_schema())
+        deleted_record = {field_: None for field_ in record_fields}
+        for id_ in stream_slice.get("deleted_ids", []):
+            deleted_record[self.id_field] = id_
```

Comment:
`self.id_field` is set to `None` above. Is it ever reset to something else?

Reply:
I see that it's used below. This is a confusing way of declaring those instance variables. They should just be set in the constructor and not allowed to be `None`

Reply:
Fixed